pkg/util/taskgroup/taskgroup.go

```diff
@@ -0,0 +1,83 @@
+// Package multierrgroup provides a mix of multierr and errgroup
+// See documentation for https://pkg.go.dev/go.uber.org/multierr and https://pkg.go.dev/golang.org/x/sync/errgroup
+package taskgroup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// Group manages goroutines and collect all the errors.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#group for more information
+type Group interface {
+	WithContext(ctx context.Context) context.Context
+	Wait() error
+	Go(name string, f func(logger *zap.Logger) error)
+}
+
+type group struct {
+	cancel context.CancelFunc
+	logger *zap.Logger
+
+	wg sync.WaitGroup
+
+	errMutex sync.Mutex
+	err      error
+}
+
+// NewGroup returns a new Group.
+func NewGroup(logger *zap.Logger) Group {
+	return &group{
+		cancel: nil, // Set separately by WithContext
+		logger: logger,
+		wg:     sync.WaitGroup{},
+
+		errMutex: sync.Mutex{},
+		err:      nil,
+	}
+}
+
+// WithContext returns a new Group with a associated Context.
```

Comment:
> returns a new Group

typo? it modifies the `group` in place and returns the `Context`.

IMO the current behavior is a bit of a footgun: IMO a safer API would be to specify the context as an argument to `NewGroup`. Maybe also worthwhile to pass it through to the function passed to `(*group).Go()`

See also:

https://github.com/neondatabase/autoscaling/blob/e3b2cecb5ae0cc903785a6c8a27c60c4ba84f852/pkg/agent/runner.go#L302-L344

Reply:
Resolving in favor of a newer comment